osutil/io.go

```diff
@@ -342,3 +342,22 @@ func AtomicSymlink(target, linkPath string) error {
 	}
 	return errors.New("cannot create a temporary symlink")
 }
+
+// AtomicLink attempts to atomically create a hardlink at linkPath, referencing
+// the given target. The process creates a temporary hardlink object pointing
+// to the target, and then proceeds to rename it atomically, replacing the
+// linkPath.
+func AtomicLink(target, linkPath string) error {
+	for tries := 0; tries < maxLinkTries; tries++ {
+		tmp := linkPath + "." + randutil.RandomString(12) + "~"
+		if err := os.Link(target, tmp); err != nil {
+			if os.IsExist(err) {
+				continue
+			}
+			return err
+		}
+		defer os.Remove(tmp)
+		return AtomicRename(tmp, linkPath)
+	}
+	return errors.New("cannot create a temporary link")
+}
```

Comment:
The implementation is almost identical, so we can share the code:

```suggestion
func atomicLinkOp(target, linkPath string, op func(target, tmp string) error, kind string) error {
	for tries := 0; tries < maxLinkTries; tries++ {
		tmp := linkPath + "." + randutil.RandomString(12) + "~"
		if err := op(target, tmp); err != nil {
			if os.IsExist(err) {
				continue
			}
			return err
		}
		defer os.Remove(tmp)
		return AtomicRename(tmp, linkPath)
	}
	return fmt.Errorf("cannot create a temporary %s", kind)
}

func AtomicSymlink(target, linkPath string) error {
	return atomicLinkOp(target, linkPath, os.Symlink, "symlink")
}

func AtomicLink(target, linkPath string) error {
	return atomicLinkOp(target, linkPath, os.Link, "link")
}
```

Reply:
Great idea, thanks!